### ISSUE TITLE
bug: Next.js app won't navigate to other pages on iOS

### ISSUE THREAD

**Issue by creativiii** [ionic-team/capacitor#3664]
### Capacitor Version

```
💊   Capacitor Doctor  💊 

Latest Dependencies:

  @capacitor/cli: 2.4.2
  @capacitor/core: 2.4.2
  @capacitor/android: 2.4.2
  @capacitor/electron: 2.4.2
  @capacitor/ios: 2.4.2

Installed Dependencies:

  @capacitor/android not installed
  @capacitor/cli 2.4.2
  @capacitor/ios 2.4.2
  @capacitor/core 2.4.2
  @capacitor/electron not installed

  Found 0 Capacitor plugins for ios:
[success] iOS looking great! 👌

```

### Platform(s)

Just iOS

### Current Behavior
Next.js uses an internal router hook to navigate to other pages. It's usually initialised like `const router = useRouter()` and then you can navigate to other pages by doing:
```
            onClick={() => {
              router.push("/test")
            }}
```

This works perfectly in both Web as well as Android, unfortunately the same can't be said about iOS. 

When trying to use the router and calling any of the [router functions to navigate ](https://nextjs.org/docs/api-reference/next/router#userouter), navigation silently fails and the page doesn't change. This happens exclusively in Capacitor and I have no idea why.

I'm aware this is also a Next.js issue, but I'm hoping there is some kind of Capacitor setting I'm missing that will allow me to fix this without refactoring all of my code.

### Code Reproduction
I created a [reproduction repository](https://github.com/creativiii/nextjs-capacitor-ios-bug). To reproduce:

- `npm install`
- `npm run build`
- `npx cap open ios`
- run the app 
- click on "Test redirect"
- Nothing happens

`npm --version` output: 6.14.8

`node --version` output: 12.19.0

`pod --version` output (iOS issues only): 1.93

**Comment by creativiii** [comment]
Upon further inspection no navigation works.

Even by just executing `window.location.href = "/test"`, the page does indeed move to /test on browser, but doesn't do anything on iOS.

So maybe this isn't a Next.js issue at all?

**Comment by creativiii** [comment]
Okay so I can't even get `<a href="/page"></a>` to work so I'm obviously missing something

**Comment by jcesarmobile** [comment]
I see a `Unhandled Promise Rejection: Error: invariant: invalid relative URL` when clicking the button, that looks like a nextjs bug to me, probably related to capacitor using capacitor scheme and that confuses the url validator (but is just a guess, I have no experience with nextjs or its router).

This `window.location.href = 'test.html'` works, but that skips using the router.

**Comment by creativiii** [comment]
> probably related to capacitor using capacitor scheme

Could you elaborate on this? What do you mean with "Capacitor scheme"?

**Comment by jcesarmobile** [comment]
Capacitor apps on iOS use a custom scheme to serve the assets.
The default scheme is capacitor, so apps are loaded from `capacitor://localhost`
nextjs router validates the relative urls and say they are not right because they don't use http or https
https://github.com/vercel/next.js/blob/764b194505caf1e3f10e58f57b8a06eaecc8ca75/packages/next/next-server/lib/router/utils/parse-relative-url.ts#L25-L30

**Comment by creativiii** [comment]
> Capacitor apps on iOS use a custom scheme to serve the assets.
> The default scheme is capacitor, so apps are loaded from `capacitor://localhost`
> nextjs router validates the relative urls and say they are not right because they don't use http or https
> https://github.com/vercel/next.js/blob/764b194505caf1e3f10e58f57b8a06eaecc8ca75/packages/next/next-server/lib/router/utils/parse-relative-url.ts#L25-L30

Is this the same `scheme` that can be changed in:
```
  "server": {
    "iosScheme": "ionic",
  },
```

Or is it something else entirely?

**Comment by jcesarmobile** [comment]
yeah, that's the scheme, but you can't set it to http nor https because those are restricted by Apple as it's reserved for remote sites.

**Comment by garrisons** [comment]
@creativiii any idea the level of difficulty for what you’re doing? I’m no capacitor expert, but I know Next pretty well.

**Comment by creativiii** [comment]
> @creativiii any idea the level of difficulty for what you’re doing? I’m no capacitor expert, but I know Next pretty well.

I'm really sorry, I originally planned to write a fix here but completely forgot. I wrote a little [article on various fixes for capacitor on ios](https://ironeko.com/posts/getting-started-with-capacitor-using-next-js-on-ios).

The gist of it, to fix the bug with Next.js' Router is as follows:

1. Install `patch-package`
2. Add a postinstall script to `package.json`

```
 "scripts": {
   "postinstall": "patch-package"
 }
```
3. Navigate to `/node_modules/next/dist/next_server/lib/router/utils/parse-relative-urls.js`, which contains the router logic that's breaking navigation.
4. The file is compiled, but luckily there's very little code. So, search for `protocol!=='http:'&&protocol!=='https:'` and replace it with `protocol!=='http:'&&protocol!=='https:'&&protocol!=='capacitor:'`
5. Run `npx patch-package next`
6. This should create a folder at `/patches`
7. If you run `npm install` you should see the output:
```
> patch-package

patch-package 6.2.2
Applying patches...
next@9.5.5 ✔
```

You can then go back and check the file you modified to make sure your changes are still there.

This fixed it for me and we've been using in production for a month or so without any problems! Let me know if you have any questions :)

**Comment by jcesarmobile** [comment]
have you considered updating next.js?

supposedly was fixed on 10.0.2 on [this PR](https://github.com/vercel/next.js/pull/16650)

https://github.com/vercel/next.js/releases/tag/v10.0.2

**Comment by creativiii** [comment]
> have you considered updating next.js?
> 
> supposedly was fixed on 10.0.2 on [this PR](https://github.com/vercel/next.js/pull/16650)
> 
> https://github.com/vercel/next.js/releases/tag/v10.0.2

Oh look at that, last time I checked they had discarded the PR.
Good to know!
